rfcs/37-dynamic_query.md

```diff
@@ -0,0 +1,168 @@
+makes it clear that the primary audience of dynamic queries is plugin authors. The interface has been designed with 
+ergonomic builders, but it's unlikely that the intended consumers of the dynamic query API will ever user it in this 
+fashion.
+
+## Rationale and alternatives
+
+While the solution to the dynamic query problem presented in this RFC does not necessarily represent the best 
+possible implementation of the internal dynamic query mechanics, it offers an ergonomic interface which can meet the
+needs of any consumer of the dynamic query API, while requiring as few changes as possible to the implementation of 
+`QueryState` and `QueryIter`, keeping the maintenance burden of offering the dynamic query interface as small as
+possible. 
+
+It's likely that implementing dynamic queries in this way will not have optimal performance since it's
+forcing a type-erased interface into a strongly typed hole, but the much lower labour requirements for implementation
+and continued upkeep justify this downside. In addition, it isn't strictly necessary to achieve the best possible
+performance. In the cases where dynamic queries are required, it's extremely unlikely for the query performance to 
+be the bottleneck when compared with the performance overhead of scripting interop.
+
+Enabling scripting is extremely important for Bevy though, even as a second class citizen within the
+ecosystem. It is inevitable that as the engine grows, games developed with the engine will find the need to include
+robust modding support into their titles, which would be dramatically more difficult without support directly
+integrated into the engine. Additionally, Bevy projects which reach a certain size will likely require some method
+of game programming for less technical designers and artists which would find Rust concepts intimidating to learn.
+Dynamic queries will enable future engine plugins to accommodate these users in an officially supported manner
+which can be reused across many different scripting solutions.
+
+
+## Prior art
+
+Back when bevy's ECS was still based on hecs, zicklag took a swing at implementing both dynamic queries and
+components in bevy PR [#632](https://github.com/bevyengine/bevy/pull/623). The underlying ECS has changed
+radically since then, so the only element which could be compared would be the interfaces for building dynamic
+queries. In that one small way, the builder pattern employed by this RFC offers a more ergonomic method for
+constructing queries, though the merits of those ergonomics for a feature only itended for advanced usage is
+questionable.
+
+## Unresolved questions
+
+These questions remain unresolved:
+- How much of the type's information should the `DynamicComponentReference` types carry with them (Layout?)?
+- How can change tracking information be captured in dynamic queries?
+- Should dynamic queries be hidden behind a feature?
```

Comment:
Given that many of the other bits of special-case functionality (e.g. audio), I think it should be reasonable to do so.

If it is, this should not be a default feature. Ideally it's named in such a way that we can easily add other related bits of functionality to the same feature flag.

This would help reduce compile times slightly (I think), but more importantly, make it much more challenging for beginners to accidentally stumble across.